### PR TITLE
Added example demonstrating creation of MNE objects from numpy arrays.

### PR DESCRIPTION
**Pull request by jaeilepp**
Issue #1382. Now it uses just sin and cos for the data. Should it be random data or something else? Is plotting the right way to demonstrate functionality?

### CODE REVIEW

**Comment by mainakjas** [comment]
you should start your filename with `plot_` if you want it to be included in the example gallery.

File: examples/create_objects_from_arrays.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+"""
+=========================================
+Creating MNE objects from data arrays
+=========================================
```
Comment (mainakjas) [inline examples/create_objects_from_arrays.py]: The equal to sign should not overshoot the text. Look at the other examples.

**Comment by mainakjas** [comment]
I would rather use the data from the sample dataset and use it. But let's see what the others think

**Comment by mainakjas** [comment]
You need to check your document using a flake8 tool

**Comment by Eric89GXL** [comment]
I actually think it's better if the data does *not* come from a sample dataset -- using `np.random` should be a good way to go. We want people to know they can put in anything they want...

File: examples/create_objects_from_arrays.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+#######################################################################
+# EvokedArray
+
+evoked_data = np.mean(epochs_data, axis=0)
+nave = len(epochs_data[0]) # Number of averaged epochs
+
+evokeds = mne.EvokedArray(evoked_data, info=info, tmin= -0.2, 
+                          comment='Arbitrary', nave=nave)
+evokeds.plot(picks=picks)
```
Comment (dengemann) [inline examples/create_objects_from_arrays.py]: last empty line missing

File: examples/create_objects_from_arrays.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+               [cos[:700], cos[1000:1700]]] 
+               
+info = mne.create_info(ch_names=['sin', 'cos'], sfreq=sfreq)
+
+epochs = mne.EpochsArray(epochs_data, info=info, events=events, 
+                         event_id={'arbitrary':1})
+
+picks = mne.pick_types(info, meg=False, eeg=False, misc=True)
+epochs.plot(picks=picks)
+plt.show()
```
Comment (dengemann) [inline examples/create_objects_from_arrays.py]: show not necessary

File: examples/create_objects_from_arrays.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+info = mne.create_info(ch_names=ch_names, sfreq=sfreq) 
+
+raw = RawArray(data, info)
+
+# Scaling of the figure. 
+# For actual EEG/MEG data a different scaling factor should be used.
+scalings = {'misc':2} 
+
+raw.plot(scalings=scalings)
+plt.show()
```
Comment (dengemann) [inline examples/create_objects_from_arrays.py]: show not necessary

**Comment by agramfort** [comment]
+1

you can use scipy.signal gaussian pulse for example:

http://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.gausspulse.html

**Comment by dengemann** [comment]
> I actually think it's better if the data does not come from a sample dataset -- using np.random should be a good way to go. 

I agree. But maybe we can use a more simple example.

**Comment by mainakjas** [comment]
I feel that the raw plot is not meaningful if you are using arbitrary data. What do the channel names mean?

File: examples/create_objects_from_arrays.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+# It is also possible to use info from another raw object.
+info = mne.create_info(ch_names=ch_names, sfreq=sfreq) 
+
+raw = RawArray(data, info)
+
+# Scaling of the figure. 
+# For actual EEG/MEG data a different scaling factor should be used.
+scalings = {'misc':2} 
+
+raw.plot(scalings=scalings)
```
Comment (mainakjas) [inline examples/create_objects_from_arrays.py]: maybe you can add a title :)

**Comment by Eric89GXL** [comment]
How about data with a few different channels with pulled from different random distributions, e.g. uniform, gaussian, etc.? Sounds like a neat example.

**Comment by dengemann** [comment]
It should be simple and direct the attention on how to create such objects, not the data.

**Comment by mainakjas** [comment]
I feel we should address the most common use case for this which is when people load data in some format that MNE does not support and construct MNE objects from it.

**Comment by dengemann** [comment]
Maybe then some of the NEO files / file formats?

**Comment by Eric89GXL** [comment]
good idea -- one more thing for @mluessi to need to install on the build bot :)

But seriously, I think an example loading data with NEO would be great.

**Comment by mainakjas** [comment]
+1 for a NEO file!

**Comment by mainakjas** [comment]
@jaeilepp : https://github.com/NeuralEnsemble/python-neo for your reference :)

**Comment by dengemann** [comment]
FYI: https://github.com/NeuralEnsemble/python-neo/issues/108

**Comment by jaeilepp** [comment]
So should there be another example for NEO? I have a feeling that just a NEO example won't reduce the confusion with the users. Or am I the only one unaware of this NEO?

**Comment by Eric89GXL** [comment]
No, just change this one example to use NEO and real data instead of fake data.

The idea is that we want people to know that they can use arbitrary other data sources by creating them from Numpy arrays. NEO is a good python library for reading file formats, so an example that uses it in together with RawArray or EvokedArray would make it clear how users can extend MNE's functionality. Make sense?

**Comment by jaeilepp** [comment]
So can I add a NEO sample data file to the sample data set somehow? Or should I create NEO files from the existing sample data and then extract the data and convert it back to MNE objects? I'm confused...

**Comment by agramfort** [comment]
don't add data to the repo. You can use an edf file from eegbci
datasets. Neo can read EDF data.

**Comment by jaeilepp** [comment]
Got it! Thanks.

**Comment by jaeilepp** [comment]
Now that I look at it, it seems that EDF is not on the list of supported formats list (https://pythonhosted.org/neo/io.html#list-of-io). I was able to get the data in NEO reader with ExampleIO, but the channel names don't transfer. I can't seem to read the matlab file with NEO either. Any suggestions?

**Comment by jaeilepp** [comment]
I could of course set the channel names by hand, but I think it wouldn't be as neat as an example. Should I do that?

**Comment by agramfort** [comment]
looking at the example. I like it much. it's simple and gets the message clearly.
I would use a demo with NEO in another example.

how do the plots look like?

**Comment by jaeilepp** [comment]
Well, the plots look a bit rough, since there are so little data for epoching and averaging, but I think they get the message through. I could add some titles and such to make it look better.
I've played a bit with the NEO libraries, and I could add a simple demo on extracting raw data from the NEOs ExampleIO object. The eventarrays in it are just oscillating signals, which seems weird to me, since it should be points of time, rather than data, if I've understood correctly.

**Comment by agramfort** [comment]
let us know when to review and don't hesitate to share screen shots

**Comment by jaeilepp** [comment]
The plots could be prettier, but I wanted to use the object's own plotting function to demonstrate, that they actually are the objects everyone knows.

File: examples/plot_objects_from_arrays.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+sfreq = 1000  # Sampling frequency
+times = np.arange(0, 10, 0.001)  # Use 10000 samples (10s)
+
+sin = np.sin(times * 10)  # Multiplied by 10 for shorter phase
+cos = np.cos(times * 10)
+sinX2 = sin * 2
+cosX2 = cos * 2
+
+data = np.array([sin, cos, sinX2, cosX2])  # Matrix of size 4 X 10000
```
Comment (agramfort) [inline examples/plot_objects_from_arrays.py]: Matrix -> Array

File: examples/plot_objects_from_arrays.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+ch_names = ['sin', 'cos', 'sinX2', 'cosX2']
+
+# Creation of info dictionary.
+# It is also possible to use info from another raw object.
+info = mne.create_info(ch_names=ch_names, sfreq=sfreq)
```
Comment (agramfort) [inline examples/plot_objects_from_arrays.py]: can you explicit channel types?

**Comment by jaeilepp** [comment]
This is ready for review, if everyone is okay with the example cases...

File: examples/plot_objects_from_arrays.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+seg = bl.segments[0]
+title = seg.file_origin
+
+data = []
+for asig in seg.analogsignals:
+    asig = asig.rescale('V').magnitude
+    data.append(asig)
+
+# data is now a stored in an array of 4 channels.
+ch_names = ['1', '2', '3', '4']
```
Comment (agramfort) [inline examples/plot_objects_from_arrays.py]: neo does not provide the channel names?

**Comment by jaeilepp** [comment]
In this particular case, no. I could use channel indices though. Let me fix it...

**Comment by dengemann** [comment]
@jaeilepp how ist going in this end. Keep in mind that we need to merge some contributions for the GSOC application.

**Comment by jaeilepp** [comment]
This is ready for merging from my part. I should have probably notified earlier. I'm not too familiar with the Github workflow yet.

**Comment by dengemann** [comment]
> This is ready for merging from my part. 

Ahh! Good to know :)

File: examples/plot_objects_from_arrays.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+#
+# License: BSD (3-clause)
+
+import mne
+
+import numpy as np
+
+print(__doc__)
+
+sfreq = 1000  # Sampling frequency
```
Comment (dengemann) [inline examples/plot_objects_from_arrays.py]: put a 79 * # separator there
and say: 
create arbitrary data

File: examples/plot_objects_from_arrays.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+cosX2 = cos * 2
+
+# Numpy array of size 4 X 10000.
+data = np.array([sin, cos, sinX2, cosX2])
+
+# Definition of channel types and names.
+ch_types = ['mag', 'mag', 'grad', 'grad']
+ch_names = ['sin', 'cos', 'sinX2', 'cosX2']
+
+# Creation of info dictionary.
```
Comment (dengemann) [inline examples/plot_objects_from_arrays.py]: put 79 # characters on the line above

File: examples/plot_objects_from_arrays.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+evokeds = mne.EvokedArray(evoked_data, info=info, tmin=-0.2,
+                          comment='Arbitrary', nave=nave)
+evokeds.plot(picks=picks, show=True, units={'mag': '-'},
+             titles={'mag': 'sin and cos averaged'})
+
+
+###############################################################################
+# Extracting data from NEO file
+
+import neo
```
Comment (dengemann) [inline examples/plot_objects_from_arrays.py]: this must go up. the most recent pep8 style guides expect all imports to happen before any other code is executed.

**Comment by dengemann** [comment]
What did we decide? Splitting this example or having NEO and the array stuff in one example?

**Comment by agramfort** [comment]
One example is fine with me.

**Comment by agramfort** [comment]
I am wondering if we should be IO examples in a separate folder in
examples...

**Comment by Eric89GXL** [comment]
Fine with me as well, and the example runs well over here after pip installing neo

**Comment by agramfort** [comment]
I'll let you merge if you're happy

File: examples/plot_objects_from_arrays.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+NEO file format is used as a source for the data.
+"""
+# Author: Jaakko Leppakangas <jaeilepp@student.jyu.fi>
+#
+# License: BSD (3-clause)
+
+import mne
+
+import neo
+import numpy as np
```
Comment (dengemann) [inline examples/plot_objects_from_arrays.py]: @jaeilepp we tend to import general libraries as numpy first, ....,  and mne last.

File: examples/plot_objects_from_arrays.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+# For actual data and different file formats, consult the NEO documentation.
+reader = neo.io.ExampleIO('fakedata.nof')
+bl = reader.read(cascade=True, lazy=False)[0]
+
+# Get data from first (and only) segment
+seg = bl.segments[0]
+title = seg.file_origin
+
+ch_names = []
+data = []
```
Comment (dengemann) [inline examples/plot_objects_from_arrays.py]: for non python new comers it's more explicit to use the buitin functions:
list()

**Comment by dengemann** [comment]
@jaeilepp almost there :-)

**Comment by dengemann** [comment]
@jaeilepp feel free to list your contribution in our whats_new.rst file . Just copy the patterns you find there.

**Comment by jaeilepp** [comment]
No worries. It's the only way I'll learn :)

File: examples/plot_objects_from_arrays.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+event_id = 1
+events = np.array([[200, 0, event_id],
+                   [1200, 0, event_id],
+                   [2000, 0, event_id]])  # List of three arbitrary events
+
+# Here a data set of 700 ms epochs from 2 channels is
+# created from sin and cos data.
+# Any data in shape (n_epochs, n_channels, n_times) can be used.
+epochs_data = [[sin[:700],      cos[:700]],
```
Comment (agramfort) [inline examples/plot_objects_from_arrays.py]: pep8

only 1 space after ,

**Comment by agramfort** [comment]
merged by rebase after addressing the nitpick

thanks @jaeilepp !

File: doc/source/whats_new.rst [hunk 15]
```diff
@@ -83,6 +83,8 @@ Changelog
 
    - Add reader function for Raw FIF files by `Teon Brooks`_
 
+   - Add example of creating MNE objects from arbitrary data and NEO files by 'Jaakko Leppakangas'_
```
Comment (mainakjas) [inline doc/source/whats_new.rst]: I just noticed this as I was rebasing my branch on to master. `Jaakko Leppangas` has the wrong quotes around it which will prevent the doc from building correctly.

Comment (agramfort) [inline doc/source/whats_new.rst]: push the fix to master directly

**Pull request closed without merging**